flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherGatewayServiceFactory.java

```diff
@@ -92,6 +91,7 @@ public ApplicationDispatcherGatewayServiceFactory(
 			throw new FlinkRuntimeException("Could not create the Dispatcher rpc endpoint.", e);
 		}
 
+		bootstrap.setErrorHandler(exception -> dispatcher.getShutDownFuture().completeExceptionally(exception));
```

Comment:
You can't pass it in the constructor because of the chicken-and-egg problem, right? You need the bootstrap to create the dispatcher but you need the dispatcher to create the `FatalErrorHandler`.

Reply:
We could also instantiate a separate error handler, right? Is the assumption that any error which `bootstrap` receives is a fatal exception and continuing the work does not make sense? If yes, then we could simply use a `FatalErrorHandler` which terminates the process.

Reply:
I agree with @aljoscha. The solution I gave was to set the error handler using a setter, but it is not elegant as a construct because there must be a default handler that does nothing, and the user has to keep in mind to set it when needed. Errors that lead to job execution termination are fatal and continuing to work does not really make sense. The semantics are that the cluster is created for the application only, so if any job in the application fails irrecoverably, it should kill also the cluster (potentially without deleting HA data).

Reply:
Terminating the dispatcher process will eventually bring down the cluster, right @tillrohrmann ?

Reply:
I don't understand the chicken-egg problem here. Why don't we pass in a bootstrap factory which creates the `ApplicationDispatcherBootstrap` with the appropriate fatal error handler when `Dispatcher.onStart` is being called?

Reply:
Yes, I think a factory is the right approach. 👌